The Usage of Valley Map data for surface T and Q observations in GSI for HRRR-based 3DRTMA

GangZhao-NOAA opened the pull request:
<!-- PLEASE READ -->
<!--
Before opening a PR, please note these guidelines:

- Each PR should only address ONE topic and have an associated issue
- No hardcoded or paths to personal directories should be present
- No temporary or backup files should be committed
- Any code that was disabled by being commented out should be removed
-->

**Description**

<!-- Please include relevant motivation and context. -->
**Motivation**
- The surface observations play very important role in RTMA/URMA. To improve the analysis of surface observations in complicated terrain area, esp. in the valley regions, (2D)RTMA/URMA has been adopting a "valley-map" technique (developed by Manuel Pondeca of NCEP/EMC) to check the valley-like terrain at the observation location and relax the observation gross error if the observation is in a valley-like area. This "valley-map" technique is useful to more efficiently use the surface observations near valley area in 2DRTMA/URMA, so it is decided to add this capability in GSI for HRRR-based 3DRTMA.

 **Method**
- The "valley-map" scheme had been implemented in GSI for 2DRTMA/URMA by Manuel years ago. But it was only set for 2DRTMA/URMA. It is needed to set up the code for HRRR-3DRTMA;
- One potential problem is how to avoid the influence of the GSD_Surface_Observation_Uselist (set in read_prepbufr)   

<!-- Please include a summary of the change and which issue is fixed. -->
**Summary of the Change in GSI**
- See this Google Doc -- [Notes on the Code Changes in GSI to Use Valley-Map data for HRRR-based 3DRTMA](https://docs.google.com/document/d/1vWTPaLBTHUZZ9ulKjQmeZ-kvgBOF3LxNK5udqnwE7HU/edit?usp=sharing)

<!-- List any dependencies that are required for this change. -->

<!-- Please provide reference to the issue this pull request is addressing. -->
<!-- For e.g. Fixes #IssueNumber -->
Fixes Issue #847 

**Type of change**

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

<!-- Please describe the tests that you ran to verify your changes and on the platforms these tests were conducted. -->
<!-- Provide instructions so we can reproduce. -->
<!-- Please also list any relevant details for your test configuration -->
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published
- [x] The modified GSI code had been tested in real-data case successfully. The observation diagnostics files for surface T obs and Q obs show that the usage indices of the observations at valley-like area are added by 0.25, which is as expected. See the following example (quoted from a obs-diag file for surface T obs)
  t KGEV     B7Hv     B7Hv     187      -0.17   36.43  278.58           908.95142           969.00000   -1    0.28936E+03    0.16954E+00        0.25000E+00     **_0.250_**  1.00
  t GDCN7    RAWS     EMPTY    195      -0.18   35.77  277.96           953.00000           524.00000   -1    0.29375E+03   -0.11074E+01        0.10000E+10  **_5000.250_**  1.00
  t POPN7    RAWS     EMPTY    195      -0.25   36.05  277.62           887.90002          1108.00000    1    0.29035E+03   -0.10170E+01        0.11000E+11     **_0.250_**  1.00
  t 0586W__a MesoWest WEATHERS 188      -0.50   36.21  278.31           797.70001          1003.00000   -1    0.29085E+03    0.20017E+01        0.10000E+11   **_102.250_**  1.00
- [x] Regression tests on WCOSS2 (Cactus) : my updated GSI commit [4d102ef](https://github.com/GangZhao-NOAA/GSI/commit/4d102ef2ba468e1194e449913cb34837b5d592f5) vs control/original GSI code (commit [#c7817c3](https://github.com/NOAA-EMC/GSI/commit/c7817c3c54dafd54c327aa3510c97ba774d70b85))
Here is the reports of the regression tests on WCOSS2 (Dogwood):
~~~
[gang.zhao@dlogin03:build] (feature/valleymap)$ ctest -N
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_rdasens
  Test #4: hafs_4denvar_glbens
  Test #5: hafs_3denvar_hybens
  Test #6: global_enkf

Total Tests: 6
[gang.zhao@dlogin03:build] (feature/valleymap)$ ctest -j 6
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  736.70 sec
2/6 Test #2: rtma .............................   Passed  977.48 sec
3/6 Test #5: hafs_3denvar_hybens ..............   Passed  1163.23 sec
4/6 Test #4: hafs_4denvar_glbens ..............   Passed  1224.50 sec
5/6 Test #6: global_enkf ......................   Passed  1431.44 sec
6/6 Test #1: global_4denvar ...................   Passed  1924.20 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 1924.27 sec
~~~

GangZhao-NOAA committed:
Added code to use valley-map data in 3D-RTMA run (l_rtma3d = .TRUE.)

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/valleymap

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/valleymap

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/valleymap

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/valleymap

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/valleymap

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/valleymap

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/valleymap

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/valleymap

GangZhao-NOAA committed:
read and use valley map data in 3DRTMA (for HRRR-CONUS 3-km domain)

GangZhao-NOAA committed:
fixed a bug (adding use valley_adjustment)

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/valleymap

GangZhao-NOAA committed:
Added grid description for 3DRTMA in CONUS domain (3-km grid)
and Akaska domain (3-km grid).

GangZhao-NOAA commented:
@ShunLiu-NOAA 
Hi Shun,
Thank you so much for assigning this PR #854 to me and sending out the invitations for review!
Could you please invite Annette Gibbs as a reviewer for this PR #854?

I created that issue last week, but I cannot assign it to myself. Please also assign the related Issue #847 to me.
Thanks again!
-Gang

ShunLiu-NOAA commented:
Assigned this issue to you.

On Wed, Mar 26, 2025 at 9:08 AM Gang Zhao ***@***.***> wrote:

> @ShunLiu-NOAA <https://github.com/ShunLiu-NOAA>
> Hi Shun,
> Thank you so much for assigning this PR #854
> <https://github.com/NOAA-EMC/GSI/pull/854> to me and sending out the
> invitations for review!
> Could you please also assign the related Issue #847
> <https://github.com/NOAA-EMC/GSI/issues/847> to me? I created that issue
> last week, but I cannot assign it to myself.
> Thanks again!
> -Gang
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/GSI/pull/854#issuecomment-2754343186>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACD5GZ3CSMWE2TCAPXUDL732WKRDFAVCNFSM6AAAAABZ2HYGZCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONJUGM2DGMJYGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: GangZhao-NOAA]*GangZhao-NOAA* left a comment (NOAA-EMC/GSI#854)
> <https://github.com/NOAA-EMC/GSI/pull/854#issuecomment-2754343186>
>
> @ShunLiu-NOAA <https://github.com/ShunLiu-NOAA>
> Hi Shun,
> Thank you so much for assigning this PR #854
> <https://github.com/NOAA-EMC/GSI/pull/854> to me and sending out the
> invitations for review!
> Could you please also assign the related Issue #847
> <https://github.com/NOAA-EMC/GSI/issues/847> to me? I created that issue
> last week, but I cannot assign it to myself.
> Thanks again!
> -Gang
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/GSI/pull/854#issuecomment-2754343186>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACD5GZ3CSMWE2TCAPXUDL732WKRDFAVCNFSM6AAAAABZ2HYGZCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONJUGM2DGMJYGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

GangZhao-NOAA committed:
Reading the height of wind sensor for 3DRTMA for provider list (same as 2DRTMA).
Updating the grid names for 3DRTMA CONUS domain and Alaska domain in subroutine ndfdgeid_info.

GangZhao-NOAA commented:
@ShunLiu-NOAA @hu5970 @guoqing-noaa @MatthewMorris-NOAA @ManuelPondeca-NOAA @RussTreadon-NOAA 
Hi all,

Following Manuel's suggestions, I made two minor modifications:

1. **read_prepbufr.f90**: line 900 ~903, adding l_rtma3d into to the same if-condition as twodvar_regional  (since 3DRTMA also needs to read the wind sensor height for height adjustment later), so 2DRTMA and 3DRTMA could share the same if-block (line 900-903), correspondingly the previously-added if-block line 905-908 for l_rtma3d is no longer needed and was removed.
2.  support_2dvar.f90: updating the grid names for 3DRTMA HRRR CONUS domain and Alaska domain (line 2147 and line 2183).

These latest changes has been committed to my working branch (feature/valleymap) of my personal fork, and this PR #854 was also updated with this latest [commit #1aea9d4](https://github.com/GangZhao-NOAA/GSI/commit/1aea9d4716821d69da4ff342523fe82abb0137de).

The GSI ctests with the latest commit are still running on WCOSS2 and Hera now. I will keep you posted.

Please make 2nd round review for this updated PR #854.

Thank you very much!

-Gang

hu5970 reviewed (approved):
Looks good.

GangZhao-NOAA commented:
@ShunLiu-NOAA @guoqing-noaa @MatthewMorris-NOAA @hu5970 @RussTreadon-NOAA
Hi All,

The ctests (regression test) with latest commit [#1aea9d4](https://github.com/GangZhao-NOAA/GSI/commit/1aea9d4716821d69da4ff342523fe82abb0137de) were done on WCOSS2 (Dogwood) and Hera. All the 6 tasks passed on WCOSS2 smoothly in one batch. The results on Hera was not that good -- 3 tasks passed and 3 
 tasks failed (global_4denvar, rrfs_3denvar_rdasens, hafs_4denvar_glbens). The good news is that there is no any GSI crash this time in Hera. The failed regression tests are mainly due to dissatisfied memory threshold or wall clock time threshold, which could happen sometimes in the regression tests. I will re-run the regression tests for those 3 tasks on hera.

Here is the result from ctests on WCOSS2/Dogwood:
~~~
[gang.zhao@dlogin03:build] (feature/valleymap)$ ctest -N
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_rdasens
  Test #4: hafs_4denvar_glbens
  Test #5: hafs_3denvar_hybens
  Test #6: global_enkf

Total Tests: 6
[gang.zhao@dlogin03:build] (feature/valleymap)$ ctest -j 6
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  735.68 sec
2/6 Test #2: rtma .............................   Passed  977.10 sec
3/6 Test #5: hafs_3denvar_hybens ..............   Passed  1105.39 sec
4/6 Test #4: hafs_4denvar_glbens ..............   Passed  1224.16 sec
5/6 Test #6: global_enkf ......................   Passed  1415.06 sec
6/6 Test #1: global_4denvar ...................   Passed  1923.44 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 1923.45 sec
~~~

Thank you!

ManuelPondeca-NOAA reviewed (approved):
The changes look good.

GangZhao-NOAA commented:
@hu5970 @ShunLiu-NOAA @guoqing-noaa @ManuelPondeca-NOAA @MatthewMorris-NOAA @RussTreadon-NOAA 
Hi All,

Here is the final result from the ctests running on Hera with the latest commit [#1aea9d4](https://github.com/GangZhao-NOAA/GSI/commit/1aea9d4716821d69da4ff342523fe82abb0137de):
In the first try of the ctest run on Hera (which is a batch run with all 6 tasks), 3 tasks passed, and other 3 tasks failed (global_4denvar, rrfs_3denvar_rdasens, hafs_4denvar_glbens). Then I re-ran ctests for each of those 3 previously failed tasks one-by-one, and they all passed.
~~~
(work-env2) [Gang.Zhao@hfe09:build] (feature/valleymap)$ ctest -N
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_rdasens
  Test #4: hafs_4denvar_glbens
  Test #5: hafs_3denvar_hybens
  Test #6: global_enkf

Total Tests: 6
(work-env2) [Gang.Zhao@hfe09:build] (feature/valleymap)$ ctest -j 6
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............***Failed  1409.94 sec
2/6 Test #6: global_enkf ......................   Passed  2056.85 sec
3/6 Test #1: global_4denvar ...................***Failed  3356.81 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  4607.57 sec
5/6 Test #4: hafs_4denvar_glbens ..............***Failed  4676.59 sec
6/6 Test #2: rtma .............................   Passed  4832.72 sec

50% tests passed, 3 tests failed out of 6

Total Test time (real) = 4832.77 sec

The following tests FAILED:
          1 - global_4denvar (Failed)
          3 - rrfs_3denvar_rdasens (Failed)
          4 - hafs_4denvar_glbens (Failed)
Errors while running CTest
Output from these tests are in: /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
(work-env2) [Gang.Zhao@hfe09:build] (feature/valleymap)$ ctest -R global_4denvar
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
1/1 Test #1: global_4denvar ...................   Passed  2641.13 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 2641.15 sec
(work-env2) [Gang.Zhao@hfe09:build] (feature/valleymap)$ ctest -R rrfs_3denvar_rdasens
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 3: rrfs_3denvar_rdasens
1/1 Test #3: rrfs_3denvar_rdasens .............   Passed  1097.25 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 1097.28 sec
(work-env2) [Gang.Zhao@hfe09:build] (feature/valleymap)$ ctest -R hafs_4denvar_glbens
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 4: hafs_4denvar_glbens
1/1 Test #4: hafs_4denvar_glbens ..............   Passed  2690.76 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 2690.78 sec
~~~

So the latest commit of my code passed the ctests on Hera as well as on WCOSS2/Dogwood.

Thank you!
